### PR TITLE
Support graceful terminator shutdown. Add more Xt ATs. Fixes GH-149

### PR DESCRIPTION
**Pull request by plorenz**

### CODE REVIEW

File: gateway/xgress_edge/listener.go [hunk 1]
```diff
@@ -296,6 +296,58 @@ func (proxy *ingressProxy) processUnbind(req *channel2.Message, ch channel2.Chan
...
+			WithField("serviceFingerprints", ns.CertFingerprints).
+			WithField("clientFingerprints", proxy.fingerprints.Prints()).
+			Error("matching fingerprint not found")
+		proxy.sendStateClosedReply("Invalid Session", req)
+		return
+	}
+
+	var cost *uint16
+	if costBytes, hasCost := req.Headers[edge.CostHeader]; hasCost {
+		updatedCost := binary.LittleEndian.Uint16(costBytes)
```
Comment (dovholuknf) [inline gateway/xgress_edge/listener.go:323]: should this just be a function ? `binary.LittleEndian` seems like the whole endian-ness stuff could be encapsulated. just a thought...

File: gateway/xgress_edge/listener.go [hunk 2]
```diff
@@ -296,6 +296,58 @@ func (proxy *ingressProxy) processUnbind(req *channel2.Message, ch channel2.Chan
...
+
+	localListener, ok := proxy.listener.factory.hostedServices.Get(token)
+
+	if ok {
+		log.Debugf("updating terminator %v to precedence %v and cost %v", localListener.terminatorId, precedence, cost)
+		if err := xgress.UpdateTerminator(proxy.listener.factory, localListener.terminatorId, cost, precedence); err != nil {
+			log.WithError(err).Error("failed to update bind")
+		}
+	} else {
+		log.Error("failed to update bind, no listener found")
```
Comment (dovholuknf) [inline gateway/xgress_edge/listener.go:347]: might just be me but from just reading the code it seems like a lot of work to get here only to find the listener isn't found? probably still have to do the check for concurrency reasons but i don't remember seeing the check earlier in the function?

File: tests/context.go [hunk 3]
```diff
@@ -724,42 +725,44 @@ func (server *testServer) acceptLoop() {
...
 
 	close(server.errorC)
-	fmt.Printf("%v: service exiting\n", server.idx)
+	pfxlog.Logger().Debugf("%v: service exiting", server.idx)
 }
 
 func (server *testServer) dispatch(conn *testServerConn) {
 	defer func() {
-		fmt.Printf("marking waiter done to conn %v-%v\n", conn.server.idx, conn.id)
+		// fmt.Printf("marking waiter done to conn %v-%v\n", conn.server.idx, conn.id)
```
Comment (dovholuknf) [inline tests/context.go:740]: leaving this in for debugging reasons?

File: tests/context.go [hunk 4]
```diff
@@ -724,42 +725,44 @@ func (server *testServer) acceptLoop() {
...
-				fmt.Printf("panic from server.dispatch: %+v\n", err)
+				log.WithError(err).Error("panic from server.dispatch")
 				server.errorC <- err
 			}
 		}
 	}()
 
 	defer func() {
-		fmt.Printf("closing conn %v-%v\n", conn.server.idx, conn.id)
+		// fmt.Printf("closing conn %v-%v\n", conn.server.idx, conn.id)
```
Comment (dovholuknf) [inline tests/context.go:757]: same comment - if there's any more i won't add any more comments :)

File: gateway/xgress_edge/listener.go (hunk 1 above)
Comment (plorenz) [reply gateway/xgress_edge/listener.go:323]: We do have some utility methods on request that I'd forgotten about for put/get*IntHeader. We don't have one for uint16, but I'll do that in a follow up PR and do some tidying here.

File: gateway/xgress_edge/listener.go (hunk 2 above)
Comment (plorenz) [reply gateway/xgress_edge/listener.go]: No reason that couldn't be at the top. Moved

File: tests/context.go (hunk 3 above)
Comment (plorenz) [reply tests/context.go]: removed

File: tests/context.go (hunk 4 above)
Comment (plorenz) [reply tests/context.go]: removed

**Pull request merged**